### ISSUE TITLE
About Instance Bank Cache

### ISSUE THREAD

**Issue by liao1fan** [HorizonRobotics/Sparse4D#120]
The function of the cache in the intance bank is, retaining the top k features and anchors with the highest confidence levels of both the historical and current features. But in your codes, the historical instance features are not been properly retained.
For a more detailed explanation, I wrote a cache demo which including the complete process of calling cache continuously for two frames, and printing the changes of key variables at each step.

**[Key output highlights in this demo]**
**First Frame**
No history cache, directly take the current frame topk.
**Second Frame**
Show how to use the historical self.confidence for fusion.
Print the confidence before and after fusion & re-select the topk process.

For easy observation, I set the parameters:
`batch_size = 1` 
`num_anchor = 5`
`num_cls = 2`
`num_temp_instances = 3`
`embed_dims = 4`
`confidence_decay=1.0`

    import torch
    class InstanceCache:

        def __init__(self, num_temp_instances=3, confidence_decay=1.0):
            self.num_temp_instances = num_temp_instances
            self.confidence_decay = confidence_decay
            self.confidence = None
            self.cached_feature = None
            self.cached_anchor = None
            self.cached_pos = None

        def cache(self, instance_feature, anchor, position, confidence):
            print("\n=== Before Cache ===")
            print("Previous cached confidence:", self.confidence)
            print("Previous cached features:", self.cached_feature)
    
            instance_feature = instance_feature.detach()
            position = position.detach()
            anchor = anchor.detach()
            confidence = confidence.detach()
    
            # Step 1: Max confidence over classes + sigmoid
            confidence = confidence.max(dim=-1).values.sigmoid()
            print("Step1 - Current frame confidence after sigmoid:", confidence)
    
            # Step 2: Fuse with history
            if self.confidence is not None:
                confidence[:, : self.num_temp_instances] = torch.maximum(
                    self.confidence * self.confidence_decay,
                    confidence[:, : self.num_temp_instances],
                )
                print("Step2 - After history fusion:", confidence)
    
            # Step 3: topk selection
            self.confidence, (self.cached_feature, self.cached_anchor, self.cached_pos) = self.topk(
                confidence, self.num_temp_instances, instance_feature, anchor, position
            )
    
            print("Step3 - Topk selected confidence:", self.confidence)
            print("Cached features after update:", self.cached_feature)
    
        def topk(self, confidence, k, *inputs):
            bs, N = confidence.shape[:2]
            confidence_topk, indices = torch.topk(confidence, k, dim=1)
            print("Topk indices:", indices)
            outputs = []
            indices_flat = (indices + torch.arange(bs)[:, None] * N).reshape(-1)
            for input in inputs:
                flattened = input.flatten(end_dim=1)
                selected = flattened[indices_flat].reshape(bs, k, -1)
                outputs.append(selected)
            return confidence_topk, outputs
    
    # ================== DEMO =====================
    
    cache_module = InstanceCache(num_temp_instances=3, confidence_decay=1.0)
    
    # First frame input
    instance_feature_1 = torch.tensor([[[0.1, 0.2, 0.3, 0.4],
                                        [0.5, 0.6, 0.7, 0.8],
                                        [0.9, 1.0, 1.1, 1.2],
                                        [1.3, 1.4, 1.5, 1.6],
                                        [1.7, 1.8, 1.9, 2.0]]])
    anchor_1 = torch.tensor([[[1]*11, [2]*11, [3]*11, [4]*11, [5]*11]])
    position_1 = torch.tensor([[[10, 20], [30, 40], [50, 60], [70, 80], [90, 100]]])
    confidence_1 = torch.tensor([[[0.1, 0.9], [0.4, 0.6], [0.3, 0.7], [0.2, 0.8], [0.9, 0.1]]])
    
    print("\n========== Frame 1 ==========")
    cache_module.cache(instance_feature_1, anchor_1, position_1, confidence_1)
    
    # Second frame input
    instance_feature_2 = torch.tensor([[[2.1, 2.2, 2.3, 2.4],
                                        [2.5, 2.6, 2.7, 2.8],
                                        [2.9, 3.0, 3.1, 3.2],
                                        [3.3, 3.4, 3.5, 3.6],
                                        [3.7, 3.8, 3.9, 4.0]]])
    anchor_2 = torch.tensor([[[6]*11, [7]*11, [8]*11, [9]*11, [10]*11]])
    position_2 = torch.tensor([[[110, 120], [130, 140], [150, 160], [170, 180], [190, 200]]])
    confidence_2 = torch.tensor([[[0.8, 0.2], [0.3, 0.7], [0.6, 0.4], [0.9, 0.1], [0.5, 0.5]]])
    
    print("\n========== Frame 2 ==========")
    cache_module.cache(instance_feature_2, anchor_2, position_2, confidence_2)

The OUTPUT is:

    ========== Frame 1 ==========
    
    === Before Cache ===
    Previous cached confidence: None
    Previous cached features: None
    Step1 - Current frame confidence after sigmoid: tensor([[0.7109, 0.6457, 0.6682, 0.6900, 0.7109]])
    Topk indices: tensor([[4, 0, 3]])
    Step3 - Topk selected confidence: tensor([[0.7109, 0.7109, 0.6900]])
    Cached features after update: tensor([[[1.7000, 1.8000, 1.9000, 2.0000],
             [0.1000, 0.2000, 0.3000, 0.4000],
             [1.3000, 1.4000, 1.5000, 1.6000]]])
    
    ========== Frame 2 ==========
    
    === Before Cache ===
    Previous cached confidence: tensor([[0.7109, 0.7109, 0.6900]])
    Previous cached features: tensor([[[1.7000, 1.8000, 1.9000, 2.0000],
             [0.1000, 0.2000, 0.3000, 0.4000],
             [1.3000, 1.4000, 1.5000, 1.6000]]])
    Step1 - Current frame confidence after sigmoid: tensor([[0.6900, 0.6682, 0.6457, 0.7109, 0.6225]])
    Step2 - After history fusion: tensor([[0.7109, 0.7109, 0.6900, 0.7109, 0.6225]])
    Topk indices: tensor([[0, 3, 1]])
    Step3 - Topk selected confidence: tensor([[0.7109, 0.7109, 0.7109]])
    Cached features after update: tensor([[[2.1000, 2.2000, 2.3000, 2.4000],
             [3.3000, 3.4000, 3.5000, 3.6000],
             [2.5000, 2.6000, 2.7000, 2.8000]]])

In Frame-2 step-3, the topk confidence is the fusion feature of the historical and current features, but you mistakenly use the index of this topk to apply to the current instance feature. Referring to the code `self.confidence, (self.cached_feature, self.cached_anchor, self.cached_pos) = self.topk(confidence, self.num_temp_instances, instance_feature, anchor, position)`,  `instance feature` and `anchor` are the *current*, not **historical and current**. So the cached features are all from the input instance feature of Frame-2.

I corrected your code, the example is as follows:

      import torch

      class InstanceCache:

        def __init__(self, num_temp_instances=3, confidence_decay=1.0):
            self.num_temp_instances = num_temp_instances
            self.confidence_decay = confidence_decay
            self.confidence = None
            self.cached_feature = None
            self.cached_anchor = None
            self.cached_pos = None
    
        def cache(self, instance_feature, anchor, position, confidence):
            print("\n=== Before Cache ===")
            print("Previous cached confidence:", self.confidence)
            print("Previous cached features:", self.cached_feature)
    
            instance_feature = instance_feature.detach()
            position = position.detach()
            anchor = anchor.detach()
            confidence = confidence.detach()
    
            # Step 1: Max confidence over classes + sigmoid
            confidence = confidence.max(dim=-1).values.sigmoid()
            print("Step1 - Current frame confidence after sigmoid:", confidence)
    
            # Step 2: Fuse with history
            if self.confidence is not None:
    
                print('cur confidence shape:', confidence.shape)
                print('history confidence shape:', self.confidence.shape)
                confidence_updated = torch.cat([confidence, self.confidence * self.confidence_decay], dim=1) # 拼接当前特征的置信度和历史特征的置信度
                instance_feature_updated = torch.cat([instance_feature, self.cached_feature], dim=1) # 拼接当前特征和历史特征
                anchor_updated = torch.cat([anchor, self.cached_anchor], dim=1) # 拼接当前anchor和历史anchor
                position_updated = torch.cat([position, self.cached_pos], dim=1) # 拼接当前位置和历史位置
    
                print("Step2 - After history fusion:", confidence_updated)
            else:
                confidence_updated = confidence
                instance_feature_updated = instance_feature
                anchor_updated = anchor
                position_updated = position
    
    
            # Step 3: topk selection
            self.confidence, (self.cached_feature, self.cached_anchor, self.cached_pos) = self.topk(
                confidence_updated, self.num_temp_instances, instance_feature_updated, anchor_updated, position_updated
            )
    
            print("Step3 - Topk selected confidence:", self.confidence)
            print("Cached features after update:", self.cached_feature)
    
        def topk(self, confidence, k, *inputs):
            bs, N = confidence.shape[:2]
            confidence_topk, indices = torch.topk(confidence, k, dim=1)
            print("Topk indices:", indices)
            outputs = []
            indices_flat = (indices + torch.arange(bs)[:, None] * N).reshape(-1)
            for input in inputs:
                flattened = input.flatten(end_dim=1)
                selected = flattened[indices_flat].reshape(bs, k, -1)
                outputs.append(selected)
            return confidence_topk, outputs
      
      # ================== DEMO =====================
      
      cache_module = InstanceCache(num_temp_instances=3, confidence_decay=1.0)
      
      # First frame input
      instance_feature_1 = torch.tensor([[[0.1, 0.2, 0.3, 0.4],
                                          [0.5, 0.6, 0.7, 0.8],
                                          [0.9, 1.0, 1.1, 1.2],
                                          [1.3, 1.4, 1.5, 1.6],
                                          [1.7, 1.8, 1.9, 2.0]]])
      anchor_1 = torch.tensor([[[1]*11, [2]*11, [3]*11, [4]*11, [5]*11]])
      position_1 = torch.tensor([[[10, 20], [30, 40], [50, 60], [70, 80], [90, 100]]])
      confidence_1 = torch.tensor([[[0.1, 0.9], [0.4, 0.6], [0.3, 0.7], [0.2, 0.8], [0.9, 0.1]]])
      
      print("\n========== Frame 1 ==========")
      cache_module.cache(instance_feature_1, anchor_1, position_1, confidence_1)
      
      # Second frame input
      instance_feature_2 = torch.tensor([[[2.1, 2.2, 2.3, 2.4],
                                          [2.5, 2.6, 2.7, 2.8],
                                          [2.9, 3.0, 3.1, 3.2],
                                          [3.3, 3.4, 3.5, 3.6],
                                          [3.7, 3.8, 3.9, 4.0]]])
      anchor_2 = torch.tensor([[[6]*11, [7]*11, [8]*11, [9]*11, [10]*11]])
      position_2 = torch.tensor([[[110, 120], [130, 140], [150, 160], [170, 180], [190, 200]]])
      confidence_2 = torch.tensor([[[0.8, 0.2], [0.3, 0.7], [0.6, 0.4], [0.9, 0.1], [0.5, 0.5]]])
      
      print("\n========== Frame 2 ==========")
      cache_module.cache(instance_feature_2, anchor_2, position_2, confidence_2)

The OUTPUT is:

    ========== Frame 1 ==========
    
    === Before Cache ===
    Previous cached confidence: None
    Previous cached features: None
    Step1 - Current frame confidence after sigmoid: tensor([[0.7109, 0.6457, 0.6682, 0.6900, 0.7109]])
    Topk indices: tensor([[4, 0, 3]])
    Step3 - Topk selected confidence: tensor([[0.7109, 0.7109, 0.6900]])
    Cached features after update: tensor([[[1.7000, 1.8000, 1.9000, 2.0000],
             [0.1000, 0.2000, 0.3000, 0.4000],
             [1.3000, 1.4000, 1.5000, 1.6000]]])
    
    ========== Frame 2 ==========
    
    === Before Cache ===
    Previous cached confidence: tensor([[0.7109, 0.7109, 0.6900]])
    Previous cached features: tensor([[[1.7000, 1.8000, 1.9000, 2.0000],
             [0.1000, 0.2000, 0.3000, 0.4000],
             [1.3000, 1.4000, 1.5000, 1.6000]]])
    Step1 - Current frame confidence after sigmoid: tensor([[0.6900, 0.6682, 0.6457, 0.7109, 0.6225]])
    cur confidence shape: torch.Size([1, 5])
    history confidence shape: torch.Size([1, 3])
    Step2 - After history fusion: tensor([[0.6900, 0.6682, 0.6457, 0.7109, 0.6225, 0.7109, 0.7109, 0.6900]])
    Topk indices: tensor([[3, 5, 6]])
    Step3 - Topk selected confidence: tensor([[0.7109, 0.7109, 0.7109]])
    Cached features after update: tensor([[[3.3000, 3.4000, 3.5000, 3.6000],
             [1.7000, 1.8000, 1.9000, 2.0000],
             [0.1000, 0.2000, 0.3000, 0.4000]]])

In corrected output, the cached features are the TRUE topk features of historical and current features. `[3.3000, 3.4000, 3.5000, 3.6000]` is from Frame-2. `[1.7000, 1.8000, 1.9000, 2.0000]` and `[0.1000, 0.2000, 0.3000, 0.4000]` are from Frame-1.

**Comment by Blake-Jiang** [comment]
Friendly note: This might already be covered in update() 😊 [update](https://github.com/HorizonRobotics/Sparse4D/blob/main/projects/mmdet3d_plugin/models/instance_bank.py#L147)
